[NEXTLEVEL 클린코드 리액트 조기문] 장바구니 미션 Step1

joel-jo-querypie opened the pull request:
안녕하세요! 
잘 부탁드립니다!! 고민들은 슬랙 고민곰윈에 계속해서 공유했던 내용입니다!
감사합니다 :) 

# 주안점
- TDD
- test를 잘 작성하려고 했으나 어렵네요 ㅜㅜ 
 
# 웹 VS CODE 환경 
[바로가기](https://github.dev/guymoon/react-shopping-cart)

# 고민 + 시도 이야기
- [🎯 TDD 시도!](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/TDD%EC%8B%9C%EB%8F%84.md)
- [🎯 redux-saga테스트 with redux-saga-test-plan](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/redux-saga-test-plan%EC%9D%84%20%EC%9D%B4%EC%9A%A9%ED%95%9C%20redux-saga%20%ED%85%8C%EC%8A%A4%ED%8A%B8!.md)
- [🎯 products 관련 action들을 다루며 은닉화 실패..
](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/products%20%EA%B4%80%EB%A0%A8%20action%EB%93%A4%EC%9D%84%20%EB%8B%A4%EB%A3%A8%EB%A9%B0%20%EC%9D%80%EB%8B%89%ED%99%94%20%EC%8B%A4%ED%8C%A8...md)
- [🎯 payload 타입 지정, 설계에 대한 고민](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/payload%20%ED%83%80%EC%9E%85%20%EC%A7%80%EC%A0%95%2C%20%EC%84%A4%EA%B3%84%EC%97%90%20%EB%8C%80%ED%95%9C%20%EA%B3%A0%EB%AF%BC.md)
- [🎯 test 할 때 주입 받는 값의 위치에 대한 고민](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/test%20%ED%95%A0%20%EB%95%8C%20%EC%A3%BC%EC%9E%85%20%EB%B0%9B%EB%8A%94%20%EA%B0%92%EC%9D%98%20%EC%9C%84%EC%B9%98.md)
- [🎯 리듀서에서 state와 action 구조 분해 할당이 가독성을 높힐 수 있을지에 대한 고민](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%EB%A6%AC%EB%93%80%EC%84%9C%EC%97%90%EC%84%9C%20state%EC%99%80%20action%20%EA%B5%AC%EC%A1%B0%20%EB%B6%84%ED%95%B4%20%ED%95%A0%EB%8B%B9%EC%9D%B4%20%EA%B0%80%EB%8F%85%EC%84%B1%EC%9D%84%20%EB%86%92%ED%9E%90%20%EC%88%98%20%EC%9E%88%EC%9D%84%EC%A7%80%3F.md)
- [🎯 어떻게하면 request에 관련된 method, url 을 효율적으로 관리할 수 있을지에 대한 고민](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%EC%96%B4%EB%96%BB%EA%B2%8C%ED%95%98%EB%A9%B4%20request%EC%97%90%20%EA%B4%80%EB%A0%A8%EB%90%9C%20method%2C%20url%20%EC%9D%84%20%ED%9A%A8%EC%9C%A8%EC%A0%81%EC%9C%BC%EB%A1%9C%20%EA%B4%80%EB%A6%AC%ED%95%A0%20%EC%88%98%20%EC%9E%88%EC%9D%84%EC%A7%80%3F.md)
- [🎯 하나의 모델과 관련된 액션들을 하나로 관리하는 것이 좋을지? 아니면 비동기 액션은 따로 관리하는 것이 좋을지?](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%ED%95%98%EB%82%98%EC%9D%98%20%EB%AA%A8%EB%8D%B8%EA%B3%BC%20%EA%B4%80%EB%A0%A8%EB%90%9C%20%EC%95%A1%EC%85%98%EB%93%A4%EC%9D%84%20%ED%95%98%EB%82%98%EB%A1%9C%20%EA%B4%80%EB%A6%AC%ED%95%98%EB%8A%94%20%EA%B2%83%EC%9D%B4%20%EC%A2%8B%EC%9D%84%EC%A7%80%3F%20%EC%95%84%EB%8B%88%EB%A9%B4%20%EB%B9%84%EB%8F%99%EA%B8%B0%20%EC%95%A1%EC%85%98%EC%9D%80%20%EB%94%B0%EB%A1%9C%20%EA%B4%80%EB%A6%AC%ED%95%98%EB%8A%94%20%EA%B2%83%EC%9D%B4%20%EC%A2%8B%EC%9D%84%EC%A7%80%3F.md)
- [🎯 함수라면 함수임을 알려주는 네이밍이어야 될까?](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%ED%95%A8%EC%88%98%EB%9D%BC%EB%A9%B4%20%ED%95%A8%EC%88%98%EC%9E%84%EC%9D%84%20%EC%95%8C%EB%A0%A4%EC%A3%BC%EB%8A%94%20%EB%84%A4%EC%9D%B4%EB%B0%8D%EC%9D%B4%EC%96%B4%EC%95%BC%20%EB%90%A0%EA%B9%8C%3F.md)

# 미션 진행하며 배운 것 + 에러 + 팁 등 기록 
- [🎯 TS 이것저것](https://www.notion.so/g1moon/43117d8c605b466eb7f3201ef0e1a65b#59102e64e7e84db49a76e38d870c340b)
- [🎯 test 환경에서 redux-logger 없애기](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/test%20%ED%99%98%EA%B2%BD%EC%97%90%EC%84%9C%20redux-logger%20%EC%97%86%EC%95%A0%EA%B8%B0.md)
- [🎯 next.js + jest + testing-library 환경설정](https://www.notion.so/g1moon/jest-next-08c42b9b86034ccba3be8ca9f28025ec)
- [🎯 useDispatch 과 useSelector 의 typed versions!](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/useDispatch%C2%A0%EA%B3%BC%C2%A0useSelector%20%20%EC%9D%98%20typed%20versions%20!.md)
- [🎯 에러핸들링](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%EC%97%90%EB%9F%AC%ED%95%B8%EB%93%A4%EB%A7%81.md)

## 📝 Requirements

### 필수 요구사항

- [x] `React Testing Library` & `Jest`를 활용해 자유로운 단위의 테스트를 진행한다.

#### GNB

- [x] `로고`를 누르면 상품목록 페이지로 이동한다.
- [x] `장바구니` 버튼을 누르면 장바구니 페이지로 이동한다.
- [x] `주문목록` 버튼을 누르면 주문목록 페이지로 이동한다.

#### 상품목록

- [x] 상품들은 n x 4 레이아웃으로 보여진다.
    - [ ] 반응형 레이아웃을 구현한다.
- [x] 상품들에는 사진, 이름, 금액이 보여진다.
- [x] 장바구니 버튼을 클릭하면 장바구니 페이지로 이동한다.
- [x] 페이징을 적용한다.

### 심화 요구사항

- [ ] 도출된 요구사항을 기반으로 `User Flow Diagram` 혹은 `Flow Chart` 작성
- [ ] UI/UX
    - [x] 사용자를 위한 로딩 환경 개선
    - [x] 페이징 혹은 인피니티 스크롤 적용 (별도의 API 없음)
        - [x] 뒤로가기 및 페이지 전환시 기존 페이지 및 스크롤 위치 기억
    - [x] 상품이 없을 때와 같은 다양한 `Edge Case` 대응
    - [ ] 반응형 레이아웃 구현
    - [ ] 별도의 모바일 레이아웃 추가 제공
    - [ ] [배민상회](https://mart.baemin.com)를 참고하여 추가 개선 사항 반영
- [ ] 매출 증대 및 마케팅을 위해 별도의 기능 구현 (별도의 API 없음)
    - [ ] 브라우저 새로고침시 모든 상태 유지
    - [ ] 흐름을 고려한 맞춤 큐레이팅 **상품 추천** 기능
    - [ ] 구매 유도를 위한 **상품 찜** 페이지
- [ ] 매출 증대 및 마케팅을 위한 별도의 도구 추가
    - [ ] Google Analytics
    - [ ] Google Tag Manager

joel-jo-querypie committed:
docs: README.md 업데이트

joel-jo-querypie committed:
feat: 초기 셋팅
- jest 초기 셋팅
- GlobalStyle 추가
- REQUIREMENT.md 추가
- yarn.lock
- yarn-error 추가
- package.json 추가
- next.config, next-env 추가

joel-jo-querypie committed:
feat: 주문목록, 장바구니, 상품상세 페이지 추가(기본 셋팅)
- feat: 주문목록(/order) 페이지 추가
- feat: 장바구니(/cart) 페이지 추가
- feat: 상품상세(/product/[productId]) 페이지 추가

joel-jo-querypie committed:
feat: redux 기본 셋팅(saga, persist, store)
- feat: sagaMiddleware 추가
- feat: redux-logger 추가
- feat: redux-persist 추가
- feat: sagaTask 설정
  - feat: sagaTask 커스텀 타입 추가
- feat: useAppDispatch(useDispatch 타이핑) 추가
- feat: rootSaga() 추가
- feat: rootReducer 추가
- feat: rootPersistConfig 추가(아직 whitelist 없음)

joel-jo-querypie committed:
feat: COLOR 상수 추가

joel-jo-querypie committed:
feat: URL 관련 상수 추가
- feat: URL (주소 관리하는 상수) 추가
- feat: REQUEST (request method와 url을 관리하는 상수) 추가

joel-jo-querypie committed:
feat: db.json fixtures 추가

joel-jo-querypie committed:
feat: shared type 추가
- feat: Redux 관련 타입 추가 (BaseRequestReducerState, RootState, AppDispatch)
- feat: Request 관련 타입 추가 (RequestMethod, RequestInfo, RequestInfoWithParams, Request, REQUEST_TYPE, ReturnedRequestInfo)
- feat: COLOR 상수 관련 타입 추가 (Color)

joel-jo-querypie committed:
feat: Spinner 컴포넌트 추가
- feat: color 변경 기능 추가
- feat: 애니메이션으로 rotation 추가
- feat: children 값으로 아래 텍스트 보여주는 기능 추가

joel-jo-querypie committed:
feat: GNB 컴포넌트 추가
- feat: 홈 로고(WOOWA SHOP) 클릭시 홈 화면(상품 목록)으로 이동하는 기능 추가
- feat: 주문목록 버튼 클릭시 주문목록 페이지로 이동하는 기능 추가
- feat: 장바구니 버튼 클릭시 장바구니 페이지로 이동하는 기능 추가

joel-jo-querypie committed:
feat: productsService(products 관련 request)
- feat: getProducts(상품 받아 오는 async-await 함수) 추가

joel-jo-querypie committed:
feat: _app.tsx 추가
- feat: redux-persist 적용(PersistGate 로 감싸기)
- feat: GlobalStyle 적용
- feat: GNB 렌더링
- feat: redux store 적용

joel-jo-querypie committed:
feat: axios 관련 작업하는 fetcher 유틸 함수 추가

joel-jo-querypie committed:
feat: ProductListContainer 컴포넌트 추가
- feat: 로딩 상태인 경우 로딩 spinner 컴포넌트 보여주는 기능 추가
- feat: products 값 받아와 ProductList 컴포넌트에게 넘겨주는 기능 추가

joel-jo-querypie committed:
feat: 상품 목록 보여주는 ProductList 컴포넌트 추가
- feat: 상품 보여주는 기눙(ProductItem 으로 상품 뿌려주는 기능) 추가(한 페이지에 12개, NUMBER_OF_PRODUCTS_ON_ONE_PAGE)
- feat: page 상태를 감시하며 마지막 페이지인지 확인하는 기능 추가
- feat: 상품 개수가 0개라면 등록된 제품이 없다고 보여주는 기능 추가
- feat: InfiniteScroll 컴포넌트 추가해 무한 스크롤 기능 추가
  - feat: 마지막 페이지라면 InfiniteScroll 컴포넌트 없애주는 기능추가

joel-jo-querypie committed:
feat: 단일 상품을 보여주는 ProductItem 컴포넌트 추가
- feat: ProductList 컴포넌트로 부터 product 정보를 받아와 보여주는 기능 추가
  -feat: 제품 이미지, 이름, 가격을 보여주는 기능 추가

joel-jo-querypie committed:
feat: 무한 스크롤 기능 적용하는 InfiniteScroll 컴포넌트 추가
- feat: 브라우저 가장 하단에 도착했는지 확인하는 returnIsBottom 함수 추가
- feat: 스크롤 시 하단인지 확인하고, 특정 작업(props로 받아온)을 하는 기능 추가
- feat: scroll 이벤트 발생시 디바운스 적용된 함수 실행하도록 하는 기능 추가

joel-jo-querypie committed:
feat: 홈 페이지('/') 에 ProductListContainer(제품 목록 보여주는) 컴포넌트 추가

joel-jo-querypie committed:
test: ProductListContainer 테스트 추가
-  describe('ProductsListContainer 가 렌더링 된다')
  - it('처음에는 로딩중임을 보인다')

joel-jo-querypie committed:
test: ProductList 테스트 추가
- describe('<ProductList/> 제품들이 보이는 컴포넌트가 렌더링 된다')

  - describe('products 가 존재 할 때(배열 안에 product 들이 있을 때)')
    - it('products 들이 화면에 보인다')

  - describe('products가 존재하지 않을 때')
    - it('products가 존재하지 않을 때 "등록된 제품이 없습니다" 가 나타난다')

joel-jo-querypie committed:
feat: useProducts(products 값을 받아오는 커스텀 훅) 추가
  - feat: reselector를 이용해 products 모델에서 값을 추출하는 기능 추가
  - feat: products 값을 받아오는 getProducts(request를 dispatch) 추가

joel-jo-querypie committed:
feat: next/router mocking 을 위한 createMockRouter 유틸 함수 추가

joel-jo-querypie committed:
feat: products 관련 비동기 액션 추가
- feat: getProducts 비동기 액션 추가

joel-jo-querypie committed:
feat: productsSaga 추가
- feat: getProductsSaga 추가

joel-jo-querypie committed:
feat: productsReducer, productsActions 추가
- productsSlice 추가
- getProducts 리듀서 추가
- ProductsReducerInitialState 추가
- ProductsReducerState interface 추가

joel-jo-querypie committed:
feat: redux 관련 유틸 함수 추가
- feat: createModelAttributeSelector 모델 속성 셀렉터 만들어주는 유틸 함수 추가
- feat: createModelAttributeObject selector를 통해 모델 속성 값 추출하는 유틸 함수 추가

joel-jo-querypie committed:
test: GNB 컴포넌트 테스트 추가
- describe('초기에 GNB에 3개의 페이지 이동 버튼이 존재한다.')
  - it('초기에 WOOWA SHOP 로고, 장바구니, 주문목록 페이지 이동 버튼이 존재한다.')

- describe('버튼을 클릭하면 다른 페이지로 이동한다.')
  - it('홈(WOOWA SHOP 버튼)을 누르면 홈 화면으로 이동한다.'
  - it('장바구니 버튼을 누르면 장바구니 화면으로 이동한다.'
  - it('주문목록 버튼을 누르면 주문목록 화면으로 이동한다.'

joel-jo-querypie committed:
test: productsSaga 테스트
- describe('productsSaga)
  - it('store의 state는 초깃값을 가지고 있다.')
  - it('getProducts success.')
  - it('getProducts failure.')
  - it('store의 state에 받아온 products 상태를 저장 할 수 있다.')

joel-jo-querypie committed:
docs: 고민, 에러핸들링, 시도해본 내용 추가

joel-jo-querypie committed:
docs: README 업데이트

joel-jo-querypie committed:
docs: README 업데이트

woobottle commented on client/jest.config.js:
```diff
@@ -0,0 +1,20 @@
+};
+
+// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
+module.exports = createJestConfig(customJestConfig);
```
next는 따로 next/jest 라이브러리를 이용해야 하군요!!
나중에 next 하게되면 기억날것 같아요

woobottle commented on client/README.md:
```diff
@@ -0,0 +1,772 @@
+- [🎯 next.js + jest + testing-library 환경설정](https://www.notion.so/g1moon/jest-next-08c42b9b86034ccba3be8ca9f28025ec)
+- [🎯 useDispatch 과 useSelector 의 typed versions!](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/useDispatch%C2%A0%EA%B3%BC%C2%A0useSelector%20%20%EC%9D%98%20typed%20versions%20!.md)
+- [🎯 에러핸들링](https://github.com/guymoon/react-shopping-cart/blob/guymoon-step1/docs/%EC%97%90%EB%9F%AC%ED%95%B8%EB%93%A4%EB%A7%81.md)
+
```
이렇게 고민이랑 배운것들 프로젝트 내에 같이 위치시키는 것 너무 좋은것 같아요! 
오늘부터 따라하겠습니다!

woobottle commented:
gitignore에는 yarn-error.log가 추가되어있는 것 같은데 
yarn-error.log 가 커밋되어 있어서 이미 cache 된것 같아요!~~

woobottle commented on client/styles/GlobalStyle.ts:
```diff
@@ -0,0 +1,38 @@
+        }
+      }
+    `
+
```
스크린 별로 font-size 다르게 주는건 놓쳤었던 부분이네요 배워갑니다!

joel-jo-querypie replied:
죄송합니다. GlobalStyle 은 다른 곳에서 퍼왔습니다... 언급해주셔서 저도 배워갑니다....ㅋㅋㅋㅋ

woobottle commented on client/styles/GlobalStyle.ts:
```diff
@@ -0,0 +1,38 @@
+      }
+      h1, h2, h3, h4, h5, h6{
+        font-family:'Maven Pro', sans-serif;
+      }
```
혹시 header 태그들만 font-family를 다르게 적용하신 이유가 궁금한데 알수 있을까요??

joel-jo-querypie replied:
죄송합니다. GlobalStyle 은 다른 곳에서 퍼왔습니다...... 저도 궁금하네요... ㅋㅋㅋ

woobottle commented on client/styles/GlobalStyle.ts:
```diff
@@ -0,0 +1,38 @@
+      ${reset}
+      * {
+        box-sizing: border-box;
+      }
```
전체 box-sizing을 아예 border-box로 설정하셨군요!!
최근에 border-box 알게 되었는데 css 작업할때 유용하더라구요!!

woobottle commented on client/store/index.ts:
```diff
@@ -0,0 +1,48 @@
+import logger from 'redux-logger';
+import { persistStore } from 'redux-persist';
+// eslint-disable-next-line import/named
+import createSagaMiddleware, { Task } from 'redux-saga';
```
해당 부분에만 eslint disable 추가해주신 이유를 알수 있을까요???!! 궁금합니다!

joel-jo-querypie replied:
아 실수네요 ㅜㅜ 수정하겠습니다!

woobottle commented on client/shared/constants/url.ts:
```diff
@@ -0,0 +1,23 @@
+  GET_CARTS: (): RequestInfo => ({ method: 'get', url: '/carts' }),
+  POST_CART: (): RequestInfo => ({ method: 'post', url: '/carts' }),
+  DELETE_CART_BY_ID: (id: string): RequestInfo => ({ method: 'delete', url: `/carts/${id}` }),
+} as const;
```
혹시 URL과 REQUEST에서 value값들이 함수로 되어있는 이유를 여쭤볼 수 있을까요??? 궁금합니다!!

joel-jo-querypie replied:
파라미터들을 이용해 url 을 만들어야 하는 경우도 있어 함수로 만들었습니다!

woobottle commented on client/shared/types/index.ts:
```diff
@@ -0,0 +1,32 @@
+  isLoading: boolean;
+  hasError: boolean;
+  error: Error | null;
+}
```
type 정의하시느라 고민이 많으셨을것 같은 생각이 드네요 
많이 배웁니다!!

joel-jo-querypie replied:
거의 처음이라 어렵네여 ㅜ

woobottle commented on client/components/GNB/__test__/GNB.test.tsx:
```diff
@@ -0,0 +1,56 @@
+    userEvent.click(orderButton());
+    expect(mockedRouter.push).toHaveBeenCalledWith(URL.ORDER());
+  });
+});
```
테스트 코드 컴포넌트 별로 상세히 작성해주셨네요!!!👍

joel-jo-querypie replied:
더 상세하게 작성해보려고 노력했는데 아직은 많이 힘드네요 ㅜㅜ

woobottle commented on client/components/GNB/GNB.tsx:
```diff
@@ -0,0 +1,54 @@
+
+const NavButton = styled.button`
+  color: ${COLOR.WHITE};
+  cursor: pointer;
```
디테일한 부분 좋은것 같습니다!!

woobottle commented on client/components/InfiniteScroll/InfiniteScroll.tsx:
```diff
@@ -0,0 +1,35 @@
+    window.addEventListener('scroll', debouncedOnScroll);
+  }, [debouncedOnScroll]);
+
+  return <></>;
```
return의 이유가 궁금해요!!

joel-jo-querypie replied:
![Screen Shot 2022-02-23 at 6 38 57 PM](https://user-images.githubusercontent.com/44131043/155293907-c182ca21-d699-421f-b36e-2394c2e38f64.jpg)

이 이유였습니다! 아무것도 보여주기는 싫고, 리턴은 해야하므로 `<></>` 를 리턴해줬는데 이 부분 꺼림직하기는 했습니다.. 아직은 방법을 잘 모르겠어 우선은 이렇게 했습니다!

woobottle replied:
혹시 tsx 파일로 작성되어있어서 그런걸까요?

joel-jo-querypie replied:
[코드](https://github.com/guymoon/react-shopping-cart/blob/62f6099b74b018c4098a307c9cbdbccb74294110/client/components/ProductList/ProductList.tsx#L40)
<- 보시면 제가 InfiniteScroll 기능을 **컴포넌트** 로 구현해서 그런게 아닐까요? jsx을 렌더링해주는 부분에 <InfiniteScroll /> 를 jsx를 리턴하는 컴포넌트로서 넣어주었는데, dom 과 관련된 것을 아무것도 리턴하지 않는다면 에러가 생기지 않을까 싶습니다!

woobottle replied:
갑자기 든 생각인데 InfiniteScrollDispatcher????

joel-jo-querypie replied:
사실 InfiniteScroll 만 고려한 컴포넌트가 아니라서요. `isBottom` 인 경우에 -> 특정한 일을 한다 같은 느낌이니까 `InfiniteScroll` 은 아예 빠져야되지 않나 싶습니다! 만약 `<Bottom workToDoWhenArrivedBottom=(alerHello) />` 이라면 바닥에 도착했을 때 `alert('hello')` 를 해주는건데 이를 설명 할 수 있는 네이밍이면 좋겠습니다ㅎㅎㅎ

joel-jo-querypie replied:
ScrollBottomObserver 어떨까요... observe scroll is bottom 이런 느낌으로 가려했는데 scroll 다음 is가 붙으면 ScrollIsBottom 으로 매우 난해한 네이밍이 되는 것 같습니다....

woobottle replied:
ScrollBottomObserver 이름 괜찮은것 같아요!!

혹시 두개의 의견이 있는데 어떤지 알고싶어요

우선 현재 컴포넌트를 return 해주는데 살짝 애매한것 같다는 개인적 생각이어서요

1. ScrollBottomObserver를 customHook으로 만들어서 bottom 에 도달했는지 여부만 리턴 그 여부를 useEffect로 관리해서 true면 다음 페이지 return???
2. ScrollBottomObserver를 wrapper 같은 느낌으로 만들어서 안에 children을 받고 render 해주는???

joel-jo-querypie replied:
오 저 2번이 좋은 것 같은데요?!! 2번으로 호다닥 수정해봐야겠습니다!

joel-jo-querypie replied:
저 그런데 지금 수정해보고 있는데 현재 이런 상태인데요
```jsx
{!isLastPage && (
        <ScrollBottomObserver workToDoWhenArrivedBottom={flipPage} msToDelayExecuteScrollHandler={200}/>
      )}
```
 wrapper로 감싸는 느낌으로 이 컴포넌트를 변경하면 isLastPage state에 따라 조건부 랜더링하는 부분이 좀 어색해질 것 같습니다. props 로 `isLastPage` 와 같은 값을 ScrollBottomObserver 컴포넌트로 props를 통해 받아 그 조건에 따라 언마운트해주는 부분을 해도 될 것 같긴한데... 뭔가 더 복잡해지는 느낌이기도 하네요 ㅜㅜ

woobottle replied:
isLastPage state에 따라 flipPage 함수의 동작 수행만 해주고 안해주고로 이해를 했는데 렌더링이 달라지는 부분이 있었군요.... 

저는 1번 방식으로 시도 해볼것 같은데 시도해보고 insight있으면 바로 공유드릴게요!!!

joel-jo-querypie replied:
저도 무슨 이유인지 모르겠는데 <></> 를 리턴해주는 건 뭔가 찝찝하네요. 어떤 사이드 이펙트가 있을지 한번 알아봐야 할 것 같아요. 1번 방법이 좋은 것 같습니다 :) 다음 스텝에서 고려해 적용해보도록하겠습니다!

woobottle commented on client/components/ProductList/__tests__/ProductList.test.tsx:
```diff
@@ -0,0 +1,39 @@
+
+      expect(result.getByText('등록된 제품이 없습니다.'));
+    });
+  });
```
보이지 않을때도 해주셨네요!! 디테일한 부분 배워갑니다!

woobottle commented on client/components/InfiniteScroll/InfiniteScroll.tsx:
```diff
@@ -0,0 +1,35 @@
+interface Props {
+  workToDoWhenArrivedBottom: () => void;
+  msToDelay?: number;
+  isLastPage: boolean;
```
isLastPage는 아직 쓰이지 않는 변수인가요??

joel-jo-querypie replied:
아 처음에는 페이지 넘기고 하는 로직을 전부 여기에 담았다가 좀 더 재사용 가능하도록 고려하다가 분리 했는데.. 왜 린터가 잡아주지 않았죠? 린트 설정을 확인해봐야겠네요

woobottle commented on client/components/ProductList/ProductList.tsx:
```diff
@@ -0,0 +1,57 @@
+  const [page, setPage] = useState(1);
+  const [isLastPage, setIsLastPage] = useState(false);
+
+  const NUMBER_OF_PRODUCTS_ON_ONE_PAGE = 12;
```
이런 변수 선언 개인적으로 너무 좋은것 같습니다!! 
설명력이 너무 좋은것 같아요!

joel-jo-querypie replied:
매직 넘버를 최소화하려고 했습니다! 지금 생각해보니 row 의 개수와 column 의 개수를 따로 지정해 조금 더 설명을 할 수 있을 것 같네요!

woobottle replied:
반응형에 따라 다르게 값을 받아올수 있으면 더 좋을것 같기도 하네요!

joel-jo-querypie replied:
반응형 요구사항 구현하면서 이 부분은 변경될 수 있겠네요~

woobottle commented on client/components/ProductList/ProductList.tsx:
```diff
@@ -0,0 +1,57 @@
+        })}
+      </ProductListContainer>
+      {!isLastPage && (
+        <InfiniteScroll workToDoWhenArrivedBottom={flipPage} isLastPage={isLastPage} />
```
InfiniteScroll의 이름 보다 더 적절한 이름이 있을것 같은데 어떤걸 추천드려야 할지는 잘 모르겠네요 ㅜㅜ
혹시 이에대해 어떻게 생각하시는지 궁금해요!!

위치와 기능에서 설명이 좀 더 잘되는 이름이 있을까 해서요

joel-jo-querypie replied:
아 그러네요 처음에는 bottom에 도착하면 페이지를 넘겨주는 일만 하던 컴포넌트였는데 조금 더 재사용 가능하게 바꿔 볼 수 있을 것 같아서 bottom 일 때 수행 할 동작을 props로 받도록 변경했습니다. 지금 상태에서는 

- bottom 을 탐지
- bottom 인 경우 특정 일을 하는 

컴포넌트로 생각 할 수 있는데요... 이에 조금 알맞은 네이밍이 뭘까요..

그냥 bottom 이라고 하고, 이렇게하면 충분히 설명 가능할까요?!
`<Bottom workToDoWhenArrivedBottom=(alerHello) />`

좋은 조언 감사드립니다 ㅎㅎ

woobottle replied:
저는 기능 구현해야 할때 보통 npm에 나와있는 비슷한 기능을 가진 라이브러리를 참조하는데요 

이전에 사용했었던 https://github.com/thebuilder/react-intersection-observer#readme
이거 한번 참고해보세요!!

저도 좋은 naming 떠오르면 바로 공유드릴게요!

joel-jo-querypie replied:
오 저도 보면서 생각해봐야겠네요 ㅎㅎ 감사합니다 :)

woobottle commented:
redux와 redux-saga쪽에서 detail한 세팅과 많은 공을 들이신걸로 알고있어요

제가 redux와 redux-saga를 이번 프로젝트를 진행하면서 처음 접하게 되어서 해당 부분에 대한 의견은 다음 pr때 드려도 괜찮을까요??
지금 당장에는 redux 쪽에서 어떤 피드백을 드려야 할지 잘 모르겠네요 ㅜㅜ 

우선 계속 공부하면서 제 프로젝트에서도 toolkit등 적용해보고 바람직한 인사이트 드릴수 있도록 할게요!

joel-jo-querypie reviewed (commented):
우병님 빠른 리뷰 감사드려요. 덕분에 많이 배우고 새로운 인사이트도 얻을 수 있었습니다! redux 리뷰는 천천히 해주세요 ㅎㅎ 감사합니다 :) 간단하게 몇 가지 수정하고, 다시 push 하겠습니다!

joel-jo-querypie committed:
refactor: ProductListContainer 에서 isLoadingProducts 값에 따라 로딩중임을 보여주는 로직에서 -> ProductList에 isLoadingProducts 값을 props로 넘겨주고, ProductList에서 로딩중임을 보여주는 로직으로 변경

  - refactor: ProductListContainer 컴포넌트에서 로딩중인 화면 보여주는 기능 제거
  - feat: ProductList 컴포넌트에서 로딩 상태에 따라 로딩화면 보여주는 기능 추가

  - test: productList.test 로딩중일때 테스트하는 부분 추가
  - refactor(test): productList.test 부분 renderProductList 렌더함수 추가하고, 렌더 함수를 이용해 테스트하는 방식으로 리팩토링
  - chore(test): productListContainer.test 설명 구체화

joel-jo-querypie committed:
refactor: InfiniteScroll.tsx 에서 ScrollBottomObserver.tsx 로 renmae
  - InfiniteScroll 외 스크롤이 bottom 일 경우 다른 일을 할 수 있는 컴포넌트로 그에 알맞은 이름으로 변경

joel-jo-querypie committed:
refactor: ScrollBottomObserver(구 InfiniteScroll 컴포넌트)의 props rename -> msToDelay 에서 msToDelayExecuteScrollHandler 로 rename
  - scroll 이벤트 핸들러의 debounce 적용 할 시간이므로 scroll 이벤트 핸들러에게 딜레이를 준다는 느낌으로 rename

joel-jo-querypie committed:
chore: eslint-disable-next-line 삭제

joel-jo-querypie committed:
chore

joel-jo-querypie committed:
chore: 실수로 initial commit에 빠진 eslint.json, prettier.json 추가

joel-jo-querypie commented:
새롭게 수정해서 push 했습니다! 확인해보시고 merge 부탁드릴게요 :) 리뷰 정말 감사했습니다!!!!

joel-jo-querypie committed:
chore: styles-example 삭제

joel-jo-querypie committed:
test: 'react-redux' mocking
- useDispatch
- useSelector

joel-jo-querypie committed:
fix: 오타(싱글 쿼터 올바르게 닫아주지 않음)로 productSage.test 깨지는 문제 해결
  - describe('productsSaga, () => {}) 에서 describe('productsSaga', () => {})

joel-jo-querypie committed:
refactor: asyncActions.ts 에서 따로 관리 되던 비동기 액션 slice.ts 에서 관리하는 방식으로 수정

  - refactor: createAsyncAction() 하기 전에 따로 객체로 액션 한번 더 만들어주는 작업 생략하고, 바로 createAsyncAction 안에 파라미터로 바로 액션 선원과 동시에 넣어주는 방식으로 수정

  - refactor: slice 에서 관리되면 reducer initial state, types.ts 에서 관리하는 방식으로 수정

joel-jo-querypie committed:
refactor: products reducer initialState 값 slice 안으로 변경
 - store/modules/products/types.ts 에서 관리하던 initialState 값 slice 안으로 이동

 - refactor(test): 외부에서 가져오던 ProductsReducerInitialState 값 테스트 내부에서 선언하는 방식으로 수정

joel-jo-querypie committed:
feat: storybook 셋팅
- main.js
- preview.js
- pckage.json
- yarn.lock

joel-jo-querypie committed:
feat: Story에 전역 스타일, 상태 적용(storybook preview.js 수정)
- feat: Story에 GlobalStyle 적용
- feat: Story에 Redux store Provider 추가

joel-jo-querypie committed:
feat: GNB Storybook 추가

joel-jo-querypie committed:
feat: ProductItem.stories.tsx 추가
- 상품 리스트에 상품 품목을 보여주는 Product Item

joel-jo-querypie committed:
feat: ProductList.stories.tsx 추가
- 상품 리스트를 보여주는 ProductList

joel-jo-querypie committed:
chore
- chore: gitignore 추가
- chore: formatter
- chore: yarn.lock

joel-jo-querypie committed:
feat: Spinner.stories 추가
- feat: Default(Blue) Spinner 스토리 추가
- feat: Red Spinner 스토리 추가
- chore: Spinner의 Props 주석 추가

joel-jo-querypie committed:
test: ProductItem test 추가
- it('상품의 title, price, image 정보를 보여준다.')

joel-jo-querypie committed:
chore: 컴포넌트의 Props에 대한 설명 추가
- ProductItem Props 주석 추가
- ProductItemList Props 주석 추가
